Extended DNS Errors a la RFC8914

wtoorop opened the pull request:
Still TODO:

- [x] Answer [EDE code 4 Forged](https://www.rfc-editor.org/rfc/rfc8914.html#name-extended-dns-error-code-4-f) for local-data answers
- [x] Fix unit tests
- [x] Configure option(s?) whether or not local-zone should result in EDE code inclusion
- [x] Write subroutine to parse packets which are REFUSED caused by an ACL to either return REFUSED with or without EDE 18
- [x] figure out how to do ACL PROHIBITED reply FORMERROR
- [x] Document how the `log-val-level:` config options influences the returned EDE option **text**
- [x] add logic for per zone EDE for RPZ and configurable rpz-do-ede
- [x] add remote-control support for the added configuration options
- [ ] configurable stale answer

For future iterations, we would like to have cached EDEs and add DNSSEC indeterminate with a configuration option,

wtoorop committed:
Add EDE EDNS0 Option code

wtoorop committed:
Simple EDE example

Commit:
add locations for EDE errors

Commit:
Fix: use EDE codes with EDNS_OPT_APPEND_EDE

+ complete EDE codes

Commit:
Return EDE_DNSSEC_BOGUS when returning bogus answers

Commit:
Show reason when returning EDE_DNSSEC_BOGUS

wcawijngaards commented on util/data/msgreply.c:
```diff
@@ -988,6 +988,35 @@ int edns_opt_append(struct edns_data* edns, struct regional* region,
+		return 0;
+	sldns_write_uint16(opt->opt_data, (uint16_t)code);
+	if (txt_len)
+		strncpy(opt->opt_data + 2, txt, txt_len);
```
```suggestion
		memmove(opt->opt_data + 2, txt, txt_len);
```
You want to always copy that length of data, so memmove. Not use the broken strncpy behaviour, or have stripped off end with zero terminator, or shorter string with uninitialised memory.

TCY16 replied:
resolved in 84da240

Commit:
add variable bogus reason

Commit:
add local anwser blocked

Commit:
Revert "add local anwser blocked"

This reverts commit b3f60db18405f4395a5402e63a930f9a897a042b.

Commit:
Fix dont echo edns0 option list ...

when refusing to answer authoritatively.
Also remove TODO comments that were already done

Commit:
EDE Blocked with local-zone refused answers

Commit:
Merge branch 'master' into features/rfc8914-ede

Commit:
set up for tpkg test

Commit:
add localzones test

Commit:
add setup of RPZ and full tests of earlier implemented EDEs

Commit:
add ede to always_refuse and always_null

Commit:
add DNSSEC indeterminate EDE and DNAME expansion test

Commit:
add mesh bogus test, possible locations for more EDE and remove superfluous todo comments

Commit:
First step towards specific EDE DNSSEC errors

Commit:
add possible EDE spots

Commit:
Answer LDNS_EDE_RRSIGS_MISSING for normal answers with missing signatures

Commit:
add routine to do EDE on ACL blocked messages

Commit:
Merge branch 'features/rfc8914-ede' of github.com:NLnetLabs/unbound into features/rfc8914-ede

Commit:
add forgotten compile error fixes from previous commit

Commit:
Merge branch 'features/rfc8914-ede' of github.com:NLnetLabs/unbound into features/rfc8914-ede

Commit:
add routine to add EDE to ACL:refused at correct location

Commit:
change strncpy to memmove at @wcawijngaards' suggestion

Commit:
process @wcawijngaards' comments

Commit:
process @wcawijngaards' comments v2

Commit:
fix CH class response

Commit:
add QDCOUNT=0 to CHAOS query in ACL

Commit:
add EDE response to autotrust_init_fail test

Commit:
add EDE response to autotrust_init_failsig test

Commit:
add EDE responses to unittests

Commit:
add DNSKEY EDE code

Commit:
add more tests

Commit:
add test setup for DNSSEC EDEs

gthess committed:
- Introduce 'ede=<info-code>' and 'all_noedns' as options in the MATCH line for
  replay test packets.

gthess committed:
Merge branch 'features/rfc8914-ede' of github.com:NLnetLabs/unbound into features/rfc8914-ede

Commit:
change unittests to match just the ede code

gthess committed:
- Fix testcode, 0 is a valid EDE INFO-CODE.

Commit:
add dnssec ede tests

Commit:
make local_data ede inclusion configurable, rewrite local_error_encode to include this logic, and remove superfluous msgparse_check_edns_in_packet function

Commit:
Merge branch 'master' into features/rfc8914-ede

Commit:
Fix merge bugs

Commit:
remove superfluous EDE left over from merge

Commit:
setup for configurable EDEs for local-zone

Commit:
add missing parsing

chantra commented on daemon/worker.c:
```diff
@@ -1460,10 +1581,23 @@ worker_handle_request(struct comm_point* c, void* arg, int error,
+					// 	EDNS_OPT_LIST_APPEND_EDE(&edns.opt_list_out, worker->scratchpad,
+					// LDNS_EDE_STALE_ANSWER, "");
+					// }
+
```
I thought I add commented about this earlier, but I don't seem to find that message again.
Is there any plan to have stale answers EDE handled from the get go of the RFC support?

chantra commented:
Not sure if I forgot to submit my previous comment a couple of weeks ago, or if it got overridden as the diff took shape.

TL;DR I am looking forward this making in into the main branch and was wondering if [EDE stale answer (code 3)](https://www.rfc-editor.org/rfc/rfc8914.html#name-extended-dns-error-code-3-s) was going to be supported from the get go as the current diff has it commented out.
Thanks @TCY16 for tackling this!

TCY16 commented:
Hi @chantra, please don't let the commented-out code dishearten you, this is very much WIP 😄 
Although I can't guarantee that EDE stale answer will be in the eventual release, we'll definitively look at this particular piece of code again!

Commit:
change do_ede to be local-zone specific and add places for more EDE codes

Commit:
add forgotten autogenerated files

Commit:
add config option for global EDE flag, local-zone specific EDE flag and default EDE code for a zone

Commit:
add error in case of incorrect string for local-zone-default-ede

Commit:
change local-zone-default-ede keywords to '-', add missing {}, and add manpage entry for local-zone-do-ede, local-zone-default-ede, and ede-local-zones

Commit:
add ede-local-zones in the manpage and update the iana_ports.inc for some reason

Commit:
finish up adding validator EDEs and other TODOs and fix tests with more specific EDE codes

Commit:
add config options to test conf, fix local-zone EDE printing logic, and fix typo in the unbound.conf

Commit:
expand ede.tdir to do validator test for DNSKEY, RRSIG and NSEC missing

Commit:
add todo for tests and fix EDE codes for DNSKEY missing

Commit:
fix DNSSEC nsec-failure test

TCY16 committed:
Merge branch 'master' into features/rfc8914-ede

Commit:
fix rpl tests

Commit:
Document how the log-val-level: config options influences the returned EDE option text

Commit:
add logic for per zone EDE for RPZ and configurable rpz-do-ede

Commit:
remove superfluous linebreaks

Commit:
add remote control options for local-zone and RPZ do_ede

Commit:
fix missing disable in remote-control local_zone do_ede and typos

Commit:
remove debug line

wtoorop committed:
Merge branch 'master' into review-rfc8914-ede

Commit:
modify tdir to ignore localzone tests
